feat(bridge): Show dialog to prevent data loss on unsaved form

ermin-muratovic opened the pull request:
## This PR
- makes sure that a user is notified when leaving the project settings form

![image](https://user-images.githubusercontent.com/6098219/148735753-7586fe52-7e3f-44e7-90fd-679c6dc5dfe0.png)

### Related Issues
Fixes #4932

Kirdock commented on bridge/client/app/_components/ktb-project-settings/ktb-project-settings.component.ts:
```diff
@@ -248,11 +252,53 @@ export class KtbProjectSettingsComponent implements OnInit, OnDestroy {
+  // @HostListener allows us to also guard against browser refresh, close, etc.
+  @HostListener('window:beforeunload', ['$event'])
+  showNotification($event?: BeforeUnloadEvent): void {
+    if (this.canDeactivate() !== true) {
```
If it's an `Observable` it's always `false`. The content of the observable is not checked here.
In my opinion, I would not even use `Observables` here

ermin-muratovic replied:
The `Observable` allows the user to decide to stay or leave the page.

Kirdock commented on bridge/client/app/_guards/pending-changes.guard.spec.ts:
```diff
@@ -0,0 +1,80 @@
+    const canDeactivate$ = <Observable<boolean>>guard.canDeactivate(mockComponent);
+    canDeactivate$.subscribe((deactivate) => {
+      // This is the real test!
+      expect(deactivate).toBeTruthy();
```
Is this even called? Because this is asynchronous I would expect that the test runs until `mockComponent.saveForm()` and finishes without waiting for the subscribe without the use of `waitForAsync`

ermin-muratovic replied:
This simulates the user deciding to leave the page.

Kirdock replied:
I'm talking about the asynchronous `subscribe`. `subscribe` would be called right after `mockComponent.saveForm();`, but after this, Jest will tell that the test succeeded, no matter what's inside the `subscribe`, because it does not wait for ongoing asynchronous tasks.
No matter if `expect(deactivate).toBeTruthy();` or `expect(deactivate).toBeFalsy();` is called inside the subscribe, the test will pass.

ermin-muratovic replied:
You are right. I added `waitForAsync`

Kirdock commented on bridge/client/app/_guards/pending-changes.guard.spec.ts:
```diff
@@ -0,0 +1,80 @@
+    const canDeactivate$ = <Observable<boolean>>guard.canDeactivate(mockComponent);
+    canDeactivate$.subscribe((deactivate) => {
+      // This is the real test!
+      expect(deactivate).toBeFalsy();
```
here same as above

ermin-muratovic replied:
This simulates the user deciding to stay on the page, which results in deactivate to be no longer an `Observable` but `false`.

Kirdock commented on bridge/client/app/_guards/pending-changes.guard.ts:
```diff
@@ -0,0 +1,24 @@
+})
+export class PendingChangesGuard implements CanDeactivate<PendingChangesComponent> {
+  canDeactivate(component: PendingChangesComponent): boolean | Observable<boolean> {
+    const hasPendingChanges = component.canDeactivate() !== true;
```
If it's an `Observable` it's always `false`.

ermin-muratovic replied:
But if no changes are applied to the Form it is set to `true`.

Kirdock replied:
Yes, because the value is `boolean` then, but on change it is set to `Observable<boolean>` and `of(true) !== true` and `of(false) !== true`. It would be the same if `false` is set instead of assigning an observable to it

ermin-muratovic replied:
Yes, but here the `hasPendingChanges` is just used to show the Notification. But the `Observable<boolean>` is also returned by the Guard, so the Route change can be triggered after e.g. saving the form or declined by the user. Without `Observable` this will not work.

Kirdock reviewed (changes requested):
- When a project is created, the notification about unsaved changes pops up and interrupts the redirect to the settings page of the created project.
- The unsaved changes dialog should have a pulse or shaking effect when the user tries to leave the page. To make it clearer what he has to do or has to look for.
- The project in the header can still be changed when there are unsaved changes. This leads to a wrong redirect after discard or save.
- "Discard changes" also redirects. Either change it to "Leave anyway" or "Discard changes and leave" or remove the redirect.

ermin-muratovic commented:
> * When a project is created, the notification about unsaved changes pops up and interrupts the redirect to the settings page of the created project.
I fixed this.
> * The unsaved changes dialog should have a pulse or shaking effect when the user tries to leave the page. To make it clearer what he has to do or has to look for.
The only way how we can add a shake effect is using following "ugly" code:
`document.querySelector('div[aria-label="Dialog for notifying about unsaved data"]').classList.add('shake');`
I added it, but do you have any other ideas?
> * The project in the header can still be changed when there are unsaved changes. This leads to a wrong redirect after discard or save.
When changing the project and choosing to leave the page the redirect works fine, but if you choose to stay the dropdown will still have the changed value. Not sure how to "revert" the click. Do you have any idea?
> * "Discard changes" also redirects. Either change it to "Leave anyway" or "Discard changes and leave" or remove the redirect.
Actually I wanted to allow the user 3 options, "Stay", "Discard changes and leave" and "Save changes". I updated this.

Kirdock reviewed (changes requested):
There are still two bugs
- If you try to leave the page during the project creation and click "save changes", nothing is done.
- If you edit the project settings (without saving) and then delete the project, it is redirected to the wrong URL. For example, deleting the "test" project redirects then to  `/project/test/settings/project`
